### PR TITLE
Add sbd stage on a running cluster

### PR DESCRIPTION
**Pull request by liangxin1300**
## 1. Add sbd on an running cluster (without any RA running)
```
# crm cluster init sbd -s /dev/sda1 -y
  Initializing SBD......done
  Restarting cluster service
  Waiting for cluster..............done
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
Then we can see `stonith-sbd (stonith:external/sbd)` and sbd.service started on each node

## 2. Add sbd on an running cluster (with any RA running)
```
# crm cluster init sbd -s /dev/sda1 -y
  Initializing SBD......done
WARNING: To start sbd.service, need to restart cluster service manually on each node
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
Then we can restart cluster service at suitable time (`stonith-sbd (stonith:external/sbd)` already configured), then we can see sbd.service started on each node

## 3. Add diskless-sbd on an running cluster (without any RA running)
```
# crm cluster init sbd -S -y
  Initializing diskless SBD......done
  Restarting cluster service
  Waiting for cluster..............done
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
Then we can see sbd.service started on each node

## 4. Add diskless-sbd on an running cluster (with any RA running)
```
# crm cluster init sbd -S -y                                                                                                                                                        
  Initializing diskless SBD......done
WARNING: To start sbd.service, need to restart cluster service manually on each node
WARNING: Then run "crm configure property stonith-enabled=true stonith-watchdog-timeout=5s" on any node
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
Restart cluster service:
```
15sp2-1:~ # crm cluster run "crm cluster restart"                                                                                                                                            
OK: [15sp2-2]
OK: [15sp2-3]
OK: [15sp2-1]
```
Check cluster status:
```
15sp2-1:~ # crm_mon -1
Cluster Summary:
  * Stack: corosync
  * Current DC: 15sp2-2 (version 2.0.4+20200616.2deceaa3a-3.3.1-2.0.4+20200616.2deceaa3a) - partition with quorum
  * Last updated: Mon Mar 29 14:27:37 2021
  * Last change:  Mon Mar 29 14:27:22 2021 by hacluster via crmd on 15sp2-2
  * 3 nodes configured
  * 1 resource instance configured

Node List:
  * Online: [ 15sp2-1 15sp2-2 15sp2-3 ]

Active Resources:
  * d   (ocf::heartbeat:Dummy):  Started 15sp2-2
```
Check sbd status:
```
15sp2-1:~ # ps -ef|grep sbd
root     24316     1  0 14:27 ?        00:00:00 sbd: inquisitor
root     24318 24316  0 14:27 ?        00:00:00 sbd: watcher: Pacemaker
root     24319 24316  0 14:27 ?        00:00:00 sbd: watcher: Cluster
root     24349 10035  0 14:27 pts/1    00:00:00 grep --color=auto sbd
```
Set stonith related properties:
```
15sp2-1:~ # crm configure property stonith-enabled=true stonith-watchdog-timeout=5s
```

## 5. Add sbd on interactive mode on an running cluster
```
# crm cluster init sbd
  
Configure SBD:
  If you have shared storage, for example a SAN or iSCSI target,
  you can use it avoid split-brain scenarios by configuring SBD.
  This requires a 1 MB partition, accessible to all nodes in the
  cluster.  The device path must be persistent and consistent
  across all nodes in the cluster, so /dev/disk/by-id/* devices
  are a good choice.  Note that all data on the partition you
  specify here will be destroyed.

Do you wish to use SBD (y/n)? y
SBD is already configured to use /dev/sda1 - overwrite (y/n)? y
  Path to storage device (e.g. /dev/disk/by-id/...), or "none" for diskless sbd, use ";" as separator for multi path []/dev/sda1
WARNING: All data on /dev/sda1 will be destroyed!
Are you sure you wish to use this device (y/n)? y
  Initializing SBD......done
  Restarting cluster service
  Waiting for cluster..............done
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
Then we can see stonith-sbd (stonith:external/sbd) and sbd.service started on each node

## 6. Validate
When some cluster nodes down:
```
# crm cluster init sbd -s /dev/sda1 -y
ERROR: cluster.init: host "15sp2-2" is unreachable:
```
When missing -s or -S option:
```
# crm cluster init sbd -y
ERROR: cluster.init: Stage sbd should specify sbd device by -s or diskless sbd by -S option
```
When already has sbd.service running
```
# crm cluster init sbd -s /dev/sda1 -y
ERROR: cluster.init: Cannot configure stage sbd: sbd.service already running!
```

## 7. Functional test
ff1e5ec86503b6e3c58954b6c1884e5f566a3f68

### CODE REVIEW

File: crmsh/bootstrap.py
Comment (liangxin1300) [inline crmsh/bootstrap.py:144]: No need to handle `ValueError` here, upper level will handle it

**Comment by liangxin1300** [comment]
@arbulu89 @gao-yan 
Could you please take a look at this? Thanks!

File: crmsh/bootstrap.py
Comment (arbulu89) [inline crmsh/bootstrap.py:522]: This method, at least with the current name is confusing. 
The method restarts the cluster but configures `sbd` too, which looks the main purpose. I think we need to find a new name.

**Review by arbulu89 — COMMENTED** [review]
@liangxin1300 
My thoughts:
1. We are getting more and more code about `sbd`. Maybe it is a good moment to have an extra python file with the `sbd` logic.
2. I guess that you looked for more options to avoid asking the user the cluster afterwards. Can you elaborate this? Why we have not tried to implement that restart in the code?

Besides this, the code looks correct.

**Comment by liangxin1300** [comment]
> @liangxin1300
> My thoughts:
> 
> 1. We are getting more and more code about `sbd`. Maybe it is a good moment to have an extra python file with the `sbd` logic.

Yeah, that is what I think https://github.com/ClusterLabs/crmsh/issues/796 :)
I can do this after this PR got merged.

> 2. I guess that you looked for more options to avoid asking the user the cluster afterwards. Can you elaborate this? Why we have not tried to implement that restart in the code?
> 

I cannot got your point here @arbulu89 

> Besides this, the code looks correct.

**Comment by arbulu89** [comment]
> > 1. I guess that you looked for more options to avoid asking the user the cluster afterwards. Can you elaborate this? Why we have not tried to implement that restart in the code?
> 
> I cannot got your point here @arbulu89

Yes.. I wrote it wrong.
As in some scenarios the new feature asks the user to restart the other cluster nodes manually, I was wondering if this action can be implemented in the code, instead of asking the user to run this.

**Comment by liangxin1300** [comment]
> > > 1. I guess that you looked for more options to avoid asking the user the cluster afterwards. Can you elaborate this? Why we have not tried to implement that restart in the code?
> > 
> > 
> > I cannot got your point here @arbulu89
> 
> Yes.. I wrote it wrong.
> As in some scenarios the new feature asks the user to restart the other cluster nodes manually, I was wondering if this action can be implemented in the code, instead of asking the user to run this.

Like https://github.com/ClusterLabs/crmsh/pull/741#issuecomment-787588665
We adopt this policy:
* When there is **no** RA running, restart cluster automatically
* When these is any RA running, restart cluster automatically might has risk, so we adopt a more conservative policy

**Comment by liangxin1300** [comment]
@gao-yan Could please review this PR when you have time?
Thank you!

File: crmsh/bootstrap.py
Comment (gao-yan) [inline crmsh/bootstrap.py:554]: No idea how the priorities of the operators `and` and `or` are handled in python like in here :-) At least the readability is not very clear.

**Review by gao-yan — COMMENTED** [review]
Good new feature! Thanks for the nice work.

File: crmsh/bootstrap.py
Comment (wenningerk) [reply crmsh/bootstrap.py:598]: > 
> In sbd sysconfig template , `SBD_WATCHDOG_TIMEOUT` is explicitly set. I'd like to make the value there correctly reflect the default for s390(x) as well at build time for the upcoming round of maintenance of sbd. But I don't think bootstrap references anything from sbd sysconfig template, right?
> 
Autogenerated sysconfig template should actually have the timeout matching to the platform.
If that doesn't work we should check why.

Comment (gao-yan) [reply crmsh/bootstrap.py:598]: It works :-) Thanks for that BTW! I mean I about to pull that into the old SLE releases.

Comment (liangxin1300) [reply crmsh/bootstrap.py:598]: @gao-yan 
So the logic on crmsh's side is:
Before setting `stonith-watchdog-timeout`, read `SBD_WATCHDOG_TIMEOUT` value from `/etc/sysconfig/sbd`, if `SBD_WATCHDOG_TIMEOUT` was set, set `stonith-watchdog-timeout` to a negative value like `-1`?

And if `SBD_WATCHDOG_TIMEOUT` is not explicitly set, set `stonith-watchdog-timeout` to be 30s for s390 and 10s for others

Am I understand correctly?

And how about the logic in sle12sp3 and 12sp4+?

Comment (gao-yan) [reply crmsh/bootstrap.py:598]: > @gao-yan
> So the logic on crmsh's side is:
> Before setting `stonith-watchdog-timeout`, read `SBD_WATCHDOG_TIMEOUT` value from `/etc/sysconfig/sbd`, if `SBD_WATCHDOG_TIMEOUT` was set, set `stonith-watchdog-timeout` to a negative value like `-1`?

If bootstrap uses sbd sysconfig template, yes. We could keep the `SBD_WATCHDOG_TIMEOUT` setting and make sure sbd sysconfig is synchronized between nodes, and set stonith-watchdog-timeout=-1.

> 
> And if `SBD_WATCHDOG_TIMEOUT` is not explicitly set, set `stonith-watchdog-timeout` to be 30s for s390 and 10s for others

Yes.

> 
> Am I understand correctly?
> 
> And how about the logic in sle12sp3 and 12sp4+?

In order to keep the behavior unchanged, for all SLE12 releases (pacemaker-1.1.x),  `stonith-watchdog-timeout` needs to be explicitly set to a positive value no matter if `SBD_WATCHDOG_TIMEOUT` is set in sbd sysconfig or not.

File: crmsh/bootstrap.py
Comment (zzhou1) [inline crmsh/bootstrap.py:517]: s/con.get/conf.get/

File: crmsh/bootstrap.py
Comment (zzhou1) [inline crmsh/bootstrap.py:520]: `os.uname().machine` for better readability

File: crmsh/bootstrap.py
Comment (zzhou1) [inline crmsh/bootstrap.py:522]:
```
else
    self._watchdog_timeout = "10s"
```

I read Yan suggests 10s for others

File: crmsh/bootstrap.py
Comment (zzhou1) [inline crmsh/bootstrap.py:544]: How about 
```
warn("To start sbd.service, need manually restart cluster service on all nodes. Try \" crm cluster run 'crm cluster restart' \" on one node  ")
```

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed, thanks!

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Hi @gao-yan ,

I created function `_determine_watchdog_timeout` to implement above logic

Please review
Thank you!

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed as `if (self.diskless_sbd and _context.cluster_is_running) or peer`

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed

File: crmsh/bootstrap.py
Comment (gao-yan) [inline crmsh/bootstrap.py:571]: Logic is kind of obscure without knowing the contexts of the callers here, but I don't know how to improve it either :-)

**Review by gao-yan — APPROVED** [review]
Only some nitpicking here, otherwise it looks good to me.  Thanks.

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py:533]: Changed
Thanks!

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py:546]: Changed as `_restart_cluster_and_configure_sbd_ra`

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py:556]: I want to change the command as `Try \"crm cluster restart --all\"` after #797 ready

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py:586]: Add some comments in this function

**Comment by liangxin1300** [comment]
Hi @gao-yan @zzhou1 @arbulu89 
Thanks for review!
I will merge this PR tomorrow if no more issues
Thanks again!

File: crmsh/bootstrap.py
Comment (zzhou1) [reply crmsh/bootstrap.py:556]: > I want to change the command as `Try \"crm cluster restart --all\"` after #797 ready

Yeah. With that, it makes sense to add a TODO code comment inline.

**Pull request merged**
